### PR TITLE
ColorCodes: Triggers on any query ending in "color <name>"

### LINKED ISSUE #3294: ColorCodes: Triggers on any query ending in "color <name>"
### Description
This IA is triggering for any query that ends in `color <name>` causing some irrelevant results.

e.g. https://duckduckgo.com/?q=symbolism+of+the+color+red&ia=answer

We need to verify that the query doesn't contain other words when the trigger is `color`

## People to notify
/cc @mintsoft 

## Get Started
- [ ] 1) Claim this issue by commenting below
- [ ] 2) Review our [Contributing Guide](https://github.com/duckduckgo/zeroclickinfo-goodies/blob/master/CONTRIBUTING.md)
- [ ] 3) [Set up your development environment](https://docs.duckduckhack.com/welcome/setup-dev-environment.html), and fork this repository
- [ ] 4) Create a Pull Request

## Resources
- Join [DuckDuckHack Slack](https://quackslack.herokuapp.com/) to ask questions
- Join the [DuckDuckHack Forum](https://forum.duckduckhack.com/) to discuss project planning and Instant Answer metrics
- Read the [DuckDuckHack Documentation](https://docs.duckduckhack.com/) for technical help

---

IA Page:  http://duck.co/ia/view/color_codes 
[Maintainer](http://docs.duckduckhack.com/maintaining/guidelines.html): @Mailkov

### PR DESCRIPTION
**Pull request by darsnack**
## Description of new Instant Answer, or changes
- Added logic to prevent triggering when excess words are involved in the sentence that are not expected

## Related Issues and Discussions
Fixes #3294 

## People to notify
@moollaza @mintsoft 

## Testing & Review
To be completed by Community Leader (or DDG Staff) when reviewing Pull Request

**Pull Request**
- [ ] Title follows correct format (Specifies Instant Answer + Purpose)
- [ ] Description contains a valid Instant Answer Page Link (e.g. https://duck.co/ia/view/my_ia)

**Instant Answer Page** (for new Instant Answers)
- [ ] Instant Answer page is correctly filled out and contains:
    - The **Title** is appropriately named and formatted
    - The **IA topics** are present and appropriate
    - The **Description** is clear and coherent
    - **Source Name** exists if applicable
    - All **Example Queries** trigger on [Beta](https://beta.duckduckgo.com/)

**Code**
- [ ] Adheres to the [DuckDuckGo Style Guide](https://docs.duckduckhack.com/resources/code-style-guide.html)
- [ ] Behaviour is appropriately tested. If improvement, tests are adequately extended.
- [ ] There is no unnecessary files in place (such as editor config files)
- [ ] There is no API keys / secrets present
- [ ] Tests are passing (run `$ duckpan test <goodie_id>`)
    - Tester should report any failures

**Ready to merge?**

- [ ] Has this IA been deployed to and tested on [beta.duckduckgo.com](https://beta.duckduckgo.com/)?
- [ ] For larger commits, has this been approved be more than one community member?
- [ ] The number of reviews is appropriate for this type of PR
- [ ] The commit message is clear, coherent and fitting

**Pull Request Review Guidelines**: https://docs.duckduckhack.com/programming-mission/pr-review.html

---

Instant Answer Page: https://duck.co/ia/view/color_codes

### CODE REVIEW

**Comment by daxtheduck** [comment]
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

File: lib/DDG/Goodie/ColorCodes.pm [hunk 1]
```diff
@@ -65,27 +65,35 @@ sub percentify {
...
+            } else {
+                $color = $q;                    # - A presumed color
+            }
         }
     }
 
     return unless $color;                   # Need a color to continue!
     $color =~ s/\sto\s//;
 
+    return unless !$filler_count;
```
Comment (moollaza) [inline lib/DDG/Goodie/ColorCodes.pm]: No need for a double negative here: `return if $filler_count;`

File: lib/DDG/Goodie/ColorCodes.pm [hunk 2]
```diff
@@ -65,27 +65,35 @@ sub percentify {
...
+    $filler_count = 0;
     foreach my $q (map { lc $_ } grep { defined $_ } @matches) {
         # $q now contains the defined normalized matches which can be:
         if (exists $types{$q}) {
             $type = $types{$q};    # - One of our types.
         } elsif ($trigger_invert{$q}) {
             $inverse = 1;          # - An inversion trigger
         } elsif (!$trigger_filler{$q}) {    # - A filler word for more natural querying
-            $color = $q;                    # - A presumed color
+            if ($q =~ /(?:^[^#][A-Z|a-z]+\s)+/) {
```
Comment (moollaza) [inline lib/DDG/Goodie/ColorCodes.pm]: This can be simplified: `$q =~ /(?:^[a-z]+\s)+/i` -- this will match a string that starts with a word followed by a space.

Can you confirm if there are any valid multi-word colour names?

Comment (darsnack) [inline lib/DDG/Goodie/ColorCodes.pm]: I think I still need `$q =~ /(?:^[^#][a-z]+\s)+/` to ignore the case where the string starts with a `#`

File: lib/DDG/Goodie/ColorCodes.pm (hunk 1 above)
Comment (darsnack) [inline lib/DDG/Goodie/ColorCodes.pm]: Good catch! I'll fix this.

File: lib/DDG/Goodie/ColorCodes.pm (hunk 2 above)
Comment (moollaza) [inline lib/DDG/Goodie/ColorCodes.pm]: If the regex is `^[a-z` then a string starting with a `#` can't get through. `^[a-z]` means "strings that start with any character between `a` and `z` 👍 

You can play around with the regex here to see that it won't let anything bad through: https://www.debuggex.com/r/dunYvhAQDY9Nli4M


---

FWIW `^[^#][a-z]` matches on a string that starts off with _anything_ that isn't `#` (meaning spaces, other symbols, letters, numbers, etc) followed by a letter, which is broader than what you want to capture.

Comment (darsnack) [inline lib/DDG/Goodie/ColorCodes.pm]: Oh duh! Thanks for the help 👍🏾

**Comment by darsnack** [comment]
@moollaza Any idea why the build is failing?

**Comment by ManrajGrover** [comment]
@darsnack Not sure why it failed. I've restarted the build and it passed. :+1:

**Comment by darsnack** [comment]
@moollaza Is this GTM?

File: lib/DDG/Goodie/ColorCodes.pm [hunk 3]
```diff
@@ -38,11 +38,11 @@ triggers query_raw => qr/^
     (?:(inverse|negative|opposite)\s+(?:of)?)?
     (?:
         (.*?)\s*(.+?)\bcolou?r(?:\s+code)?|             # handles "rgb red color code", "red rgb color code", etc
-        (.*?)\s*(.+?)\brgb(?:\s+code)?|                 # handles "red rgb code", etc
+        (.*?)\s*(.+?)\brgb(?:\s+code)?|                            # handles "red rgb code", etc
```
Comment (mintsoft) [inline lib/DDG/Goodie/ColorCodes.pm]: Can you tidy up the whitespace changes here so I can more clearly see any modifications?

File: lib/DDG/Goodie/ColorCodes.pm [hunk 4]
```diff
@@ -65,27 +65,35 @@ sub percentify {
...
+    $filler_count = 0;
     foreach my $q (map { lc $_ } grep { defined $_ } @matches) {
         # $q now contains the defined normalized matches which can be:
         if (exists $types{$q}) {
             $type = $types{$q};    # - One of our types.
         } elsif ($trigger_invert{$q}) {
             $inverse = 1;          # - An inversion trigger
         } elsif (!$trigger_filler{$q}) {    # - A filler word for more natural querying
-            $color = $q;                    # - A presumed color
+            if ($q =~ /(?:^[a-z]+\s)+/) {
```
Comment (mintsoft) [inline lib/DDG/Goodie/ColorCodes.pm]: I don't quite understand what this is doing do we basically want to bomb out if there's a word that isn't a colour?

File: lib/DDG/Goodie/ColorCodes.pm (hunk 3 above)
Comment (darsnack) [inline lib/DDG/Goodie/ColorCodes.pm]: How can I do this without another commit? If I make a commit removing the whitespaces changes, then won't it still highlight those changes?

File: lib/DDG/Goodie/ColorCodes.pm (hunk 4 above)
Comment (darsnack) [inline lib/DDG/Goodie/ColorCodes.pm]: That's basically the idea. The original regex flow was designed to capture all words except the color. So the color would be the only word that was an unknown. However, the examples in the issue point to situations where there are multiple extra words involved in a query that isn't meant to trigger the ColorCode IA. So, if we count too many extra words, then we bomb out, because the query can't be one of the queries we wanted to match (i.e. the original regex flow was only designed to match a certain set of queries all of which support only one extra word — the color).

File: lib/DDG/Goodie/ColorCodes.pm (hunk 3 above)
Comment (kirkins) [inline lib/DDG/Goodie/ColorCodes.pm]: @darsnack the "Files changed" tab shows the differences between the original code and the state of your code in the latest commit, so no it won't be a problem if you make another commit.

File: lib/DDG/Goodie/ColorCodes.pm (hunk 4 above)
Comment (mintsoft) [inline lib/DDG/Goodie/ColorCodes.pm]: RIght, I understand. I think this whole section handling the `@matches` needs tidying up, but it's reasonably large job so it's probably better doing it separately. Can you add a few more tests demonstrating the behaviour when it should and shouldn't return results? That'd make refactoring this code easier and safer.

**Comment by mintsoft** [comment]
@moollaza LGTM, this IA needs a clean up but it's probably going to be best off as a dedicated task

**Comment by moollaza** [comment]
@darsnack Thanks! This LGTM :+1:

**Comment by bsstoner** [comment]
This was released to production 🚀
